src/Formats/IonWriter.cpp

```diff
@@ -0,0 +1,306 @@
+                // todo: Should we put a border on top of this?
+                ion_buffer.resize(ion_buffer.size() << 1);
+                stream->limit = ion_buffer.data() + ion_buffer.size();
+                stream->curr = ion_buffer.data() + total_written_bytes;
+            }
+        }
+        else
+        {
+            // else it is first call
+            total_written_bytes = 0;
+            stream->limit = ion_buffer.data() + ion_buffer.size();
+            stream->curr = ion_buffer.data();
+        }
+        return IERR_OK;
+    };
+
+    // Init stream and writer
+    ION_STREAM * stream = nullptr;
+    hWRITER writer = nullptr;
+    auto err = ion_stream_open_handler_out(&ionWriteCallback, &ion_writer_callback, &stream);
+    if (err == IERR_OK)
+    {
+        err = ion_writer_open(&writer, stream, &ion_options);
+        if (err == IERR_OK)
+        {
+            ion_stream = stream;
+            ion_writer = writer;
+        }
+        else
+        {
+            ion_stream_close(stream);
+            // todo: Should we be doing anything else here?
+        }
+    }
+}
+
+IonWriter::~IonWriter()
+{
+    if (ion_writer != nullptr)
+    {
+        ion_writer_close(ion_writer);
+        ion_stream_close(ion_stream);
+    }
+}
+
+size_t IonWriter::bytesWritten() const
+{
+    return total_written_bytes;
+}
+
+size_t IonWriter::flush()
+{
+    SIZE written_bytes;
+    ion_writer_flush(ion_writer, &written_bytes);
+    return written_bytes;
+}
+
+size_t IonWriter::finish()
+{
+    SIZE written_bytes;
+    ion_writer_finish(ion_writer, &written_bytes);
+    return written_bytes;
+}
+
+void IonWriter::final()
+{
+    size_t written_bytes = finish();
+    out.write(reinterpret_cast<char *>(ion_buffer.data()), written_bytes);
+}
+
+void IonWriter::writeInt(int64_t value)
+{
+    ion_writer_write_int64(ion_writer, value);
+}
+
+void IonWriter::writeBigInt(const char * value, size_t n)
+{
+    std::vector<BYTE> big_int_buffer(n);
+    for (size_t i = 0; i < n; ++i)
+        big_int_buffer[n - i - 1] = value[i];
+    ION_INT * ion_int_ptr = nullptr;
+    ion_int_alloc(nullptr, &ion_int_ptr);
+    ion_int_from_bytes(ion_int_ptr, big_int_buffer.data(), static_cast<SIZE>(big_int_buffer.size()));
+    ion_writer_write_ion_int(ion_writer, ion_int_ptr);
+}
+
+void IonWriter::writeBigUInt(const char * value, size_t n)
+{
+    std::vector<BYTE> big_int_buffer(n);
+    for (size_t i = 0; i < n; ++i)
+        big_int_buffer[n - i - 1] = value[i];
+    ION_INT * ion_int_ptr = nullptr;
+    ion_int_alloc(nullptr, &ion_int_ptr);
+    ion_int_from_abs_bytes(ion_int_ptr, big_int_buffer.data(), static_cast<SIZE>(big_int_buffer.size()), false);
+    ion_writer_write_ion_int(ion_writer, ion_int_ptr);
+}
+
+void IonWriter::writeFloat(float value)
+{
+    ion_writer_write_float(ion_writer, value);
+}
+
+void IonWriter::writeDouble(double value)
+{
+    ion_writer_write_double(ion_writer, value);
+}
+
+void IonWriter::writeDecimal32(Int32 value)
```

Comment:
Our binary representation of `Decimal` differs from binary representation of ION `Decimal`. We don't store scale and precision in the binary representation because it's specified during type creation (`Decimal(P, S)`) and we always know them in advance. So, our binary representation is simple:
```
:) select reinterpret(toDecimal32(-42.423, 3), 'Int32')

   ┌─reinterpret(toDecimal32(-42.423, 3), 'Int32')─┐
1. │                                        -42423 │
   └───────────────────────────────────────────────┘
```

But ION decimal has different specification: https://amazon-ion.github.io/ion-docs/docs/decimal.html
We should be able to convert our decimal to ION decimal or don't support decimals at all.

Reply:
This is still relevant